### ISSUE TITLE
Capturing audio-only

### ISSUE THREAD

**Issue by jespertheend** [w3c/mediacapture-screen-share#100]
Hi, I was wondering why in https://github.com/w3c/mediacapture-screen-share/commit/c5c484213b15ead2780b1bcd822feff7e7872f1f

`MUST reject an audio-only request if no audio track is available.`
got changed to
`MUST reject audio-only requests.`

Right now it seems like all audio-only requests need to be rejected, even if the system does support audio tracks. Is this intended?

**Comment by henbos** [comment]
Yes this is intentional. Audio capture is perceived as an optional addition to screen capture, something that can exist side-by-side video but is not enforced to have to exist. A `getDisplayMedia({audio:true, video:true})` is allowed to return only a video track if the user wants to share video but not audio, or if the implementation/platform does not support audio capture. A request cannot limit the user's selection (e.g. which window to share since the application is not allowed to know which windows exists), and a request cannot force a user to pick audio. This is different from `getUserMedia()` where audio is non-optional if requested.

In the screen capture case, audio is a lot more technical and its availability to some extent platform-dependent. The spec is purposefully vague about what audio has to be provided. Even to the point of making it optional.

**Comment by henbos** [comment]
For example, I believe Mac does not allow capturing system audio, so the only audio that might be implementable on that platform is tab/browser audio.

**Comment by jespertheend** [comment]
Alright, thanks for the clarification. That makes sense. The reason why I brought this up is because in my use case I only need the system audio. Right now I’m using a chrome extension to capture both the screen and system audio. But capturing the screen seems to lower my framerate significantly. So that’s why I was hoping there would be a way to request audio only.

Though I guess this is very much an edge case and this would probably require drastic changes to the spec or maybe even a new API altogether.

**Comment by jespertheend** [comment]
Would calling `MediaStreamTrack.stop()` solve this or would that just stop the screen capture altogether? I couldn't find anything about this in the current spec.

**Comment by henbos** [comment]
I believe you can stop the video track without stopping the audio track (like getUserMedia()), which should fix your performance issues. Undeniably this is a workaround for using an API not the way it was intended. I.e. the user gets prompted to share their SCREEN, and depending on their choice/platform you MAY get audio. It's not ideal.

**Comment by youennf** [comment]
Agreed this is not ideal.
Should the privacy indicator give information about the audio being captured?
If there is a possibility to capture audio but not screen, should it be reflected there as well?

**Comment by jespertheend** [comment]
Yeah I agree. It’s not ideal but it’s the best solution for now. The alternative is asking users to install software to route system audio to a virtual microphone such as soundflowerbed. So using this APi is much more user friendly. I’ll try to stop the video track and see if that works.

**Comment by innerop** [comment]
Was googling and arrived here.

I realize this is a different API, not the Extension API, but I've given up on getting browser/tab audio from getDisplayMedia on Mac and I was trying to use tabCapture from an extension since I know the tab audio capture works great via the tabCapture API

However, ...

When using chrome.tabCapture.capture({audio: true, video: true}, (stream) => { ... }) in Chrome, it will mute the audio and so no audio is capture, almost as if it was done intentionally to prevent DRM'd and Youtube media capture

if you set video to false then tab audio is captured normally. If you set the audio to false then video is captured normally. But never video and audio at the same time.

Is screen sharing with tab audio via Chrome a dead end?

This is not about capturing system audio. Just tab audio and video at once.

**Comment by guest271314** [comment]
@henbos 

> or if the implementation/platform does not support audio capture

> audio is a lot more technical and its availability to some extent platform-dependent

Where the architecture _does_ support capturing system wide audio output does the language MAY at

> In the case of audio, the user agent _MAY_ present the end-user with audio sources to share. Which choices are available to choose from is up to the user agent, and the audio source(s) are not necessarily the same as the video source(s). An audio source may be a particular application, window, browser, the entire system audio or any combination thereof. 

signal to implementers to not expose the device that captures system-wide or specific audio devices?

In the case of *nix, PulseAudio _does_ provide a means to capture system-wide audio, however, Chromium does not expose that device at either `getUserMedia()` or `getDisplayMedia()` prompts or `enumerateDevices()`. The only workaround that have been able to use to set the device to be captured to "Monitor of <device>" is to manually change the device is recording a `MediaStream` with `MediaRecorder` and _during_ the recording changing the output device at sound settings GUI, which results in `getUserMedia()` setting the device captured to the user selected device at system GUI for all subsequent calls, which is the expected result, here, though there is no way to change the setting using an API in the browser at Chromium, in brief see https://github.com/web-platform-tests/wpt/issues/23084.

Thus, the question is does a separate issue and PR need to be filed for the MAY language, which if an issue is filed at Chrome bug tracker, could Chromium authors simply say: The specification says "MAY", therefore we do no have to implement audio _output_ capture for `getDisplayMedia()`, as that is optional, and even though possible at *nix, we have decided to _not_ implement that part of the API, as it is not required by the specification?

**Comment by bradisbell** [comment]
@henbos @jan-ivar Is it still the consensus that audio-only streams are out of scope for this API?

Audio-only use cases are artificially damaged at the moment.  If a user wants to capture audio from the system or tab, we first have to prompt them to share their screen, hope they can figure out checking the "share audio" box, then stop the video stream, and hope they don't care about the UI saying we're sharing their screen when we aren't.  From there, the audio only stream works fine on supported platforms.

The technical concerns raised a few years ago don't seem as relevant now.  It seems like the only thing holding back audio use cases is this specification, and the established position that audio is out of scope.

What are the specific concerns about `getDisplayMedia({video: false, audio: true})` today?  Is it a good time to revisit?

**Comment by henbos** [comment]
I'm not up to speed on this anymore, but IIUC system audio capture does not exist on all popular OSes. I suspect you would be less interested in audio tab capture?

**Comment by eladalon1983** [comment]
I think it's counterproductive to force a user to share their (screen AND system audio) when they only want to share (system audio). Similarly for sharing (tab pixels AND tab audio) when they only wish to share (tab audio). Let's reopen this. If there's a simple explanation for such a constraint, and that explanation still holds, then we can always re-close.

**Comment by henbos** [comment]
Privacy and security:
I'm not saying audio-only capture is necessarily better or worse from a privacy and security perspective, but it may be different enough to be worth revisting if the privacy and security versus use cases trade-off is still worth it.
- Unlike screen sharing, audio-only capture might not be as clear to the user that capturing is happening or what is being captured.
- Unlike screen sharing, audio-only capture is much more limited in use cases. I'm personally having trouble coming up with convincing ones.

But even if that is resolved: doing an API that only works on some OSes, or that works different on different OSes, seems controversial in and of itself.
- So I we should have some really strong use cases for this... what are they?

Btw I agree that doing audio+video capture when you only want audio is counterproductive - it's a workaround to not having solved the audio-only capture use case. But the question is: do we want to solve this use case?

I would start with "what use cases do we want to solve?"

**Comment by eladalon1983** [comment]
The question of use-cases is interesting and I'll get back to that. Quick recap first of the benefits of supporting audio-only, assuming use-cases exist:
* Benefit to the user:
  * Enhance privacy through not having to share more than strictly intended.
* Benefit to the application:
  * Improved relationship with the user - no need to alarm the user by asking for irrelevant permissions. (Users should not be expected to understand that the app cannot capture audio without video.)
  * Improved efficiency (avoid unnecessary CPU/GPU load by unnecessarily maintaining screen- or tab-capturing).
* Benefit to the browser:
  * Improved efficiency (see above).
* Benefit to the Web platform:
  * Better-educated users - make it more common to only share what's strictly necessary.

Clearly an all-round win - provided we have the use-cases. Back to that, then. :-)
* Audio-record meetings into a file. (One app captures another; no integration needed.)
* Record lectures, music, etc.  (One app captures another; no integration needed.)
* Call centers. Assume you're building software that exposes some functionality to the operator - some CRM-like software. Your software imports some functionality by embeding some VC application in an iframe, allowing you to talk to the customer/patient/colleague while interacting with the CRM-like software concerning his call. By recording tab audio, your own software now quickly gets to record the call to a file without having to integrate more closely with the VC software you're embedding. No need to ask them to make changes to accommodate you as a customer, wait for that to happen - you just implement recording of (i) the tab and (ii) the mic and you're all set.

**Comment by bradisbell** [comment]
My use cases involve live stream production, where users capture audio from multiple sources.  This often includes other software such as Skype, Zoom, DJ-style media players, other tabs sharing audio, etc.

Audio-only streaming is very much relevant.  Video is not always required nor desired.  Native applications have no problem in capturing audio in this way.  We only have to look to the success of software such as OBS, Wirecast, Rocket Broadcaster, Voicemeeter, etc., to understand that we need this capability on the web.  All of these support audio-only modes and capturing from other applications.

The users should be empowered.  This specification and its implementation are the limiting factors right now, as I see it.  

I propose:

 - `getDisplayMedia({audio: true, video: true})` will show the current sharing dialog, with "share audio" checkbox pre-checked.  (Users may un-check it.)
 - `getDisplayMedia({audio: true, video: false})` will show the current sharing dialog, with the correct verbiage explaining that only audio is being shared.
 - `getDisplayMedia({audio: {exact: true}, video: ...})` will show the current sharing dialog with "share audio" checkbox pre-checked and locked.  User may cancel to reject.  Only surfaces supporting audio sharing will be selectable.  If audio sharing is not available, an overconstrained error should throw.
 - Shared applications may have a similar overlay as they do now, but with verbiage explaining that only audio is being shared.

In other words... I propose that this API match what is expected and consistent with the `getUserMedia()` API.  This solves for existing use cases, providing a generic interface for future use cases, simplifying the API for application developers, hopefully resulting in minimal user agent changes, while continuing to offer a user experience that is very similar to what exists today.

**Comment by schreibmachine** [comment]
I would like to second what @r00ster91 said. I am currently trying to build a web-based audio visualizer and wanted users to be able to capture their microphone and/or system sounds. After a frustrating day trying to find out how to access system audio I now see that it's simply not (easily) possible. And now also understand why seemingly all web-based audio visualizers use uploaded media... If privacy is a concern, what is the difference of system audio to the microphone? Isn't the microphone an even more sensitive audio stream?

**Comment by chrisguttandin** [comment]
If no one wants to present it this time, I would be happy to present it in one of the next meetings, if it's okay for a CG member to present something. But I can't attend the upcoming meeting in two weeks.

**Comment by eladalon1983** [comment]
The process to [join](https://www.w3.org/community/sccg/join) community groups is a lot more lightweight than that of joining working groups, and I believe it does not involve payment to the W3C. At any rate, anyone is welcome to present.

**Comment by cybex-dev** [comment]
Coming from this [SO post](https://stackoverflow.com/a/58623841) with high hopes, though it seems audio-only capture is a far way off from a production ready feature?
